Encryption

e2r3p13 opened the pull request:
Implements encryption/decryption on server-side, using AES256/GCM:

- At enroll time, the server now generates a 256bits symmetric key, which is added to the QRcode. On connection, the new attester generates a UUID and sends it to the server. The server then seals the symmetric key to the TPM Storage Root Key (SRK), and stores the resulting blobs in two files: `/etc/ultrablue/UUID` and `/etc/ultrablue/UUID.pub`.

- During a classic attestation, the client sends the UUID it previously stored, and the server will retrieve and unseal the corresponding key.

In both cases, the first message is the client UUID, and every message after this one are encrypted.

A random IV is generated for each message, and the data send to the characteristic channel actually looks like this:
```
[size][IV][Ciphertext [CBOR [Plaintext]]]
0    4   16                           16 + size
```

This pull request also contains minor changes:
- TPM related functions are moved out from `utils.go` to `tpm.go`
- The `RegistrationData` structure has been renamed to `EnrollData`
- Adds/Updates some comments

e2r3p13 committed:
[server] Add a 256 bits key in the QR code

When enrolling a new device, a key of 32 bytes (256 bits) is
generated through the crypto/rand package, and its hex string
representation will be shown in the QR code.

Data in the QR code is now formatted in JSON, with the following shape:

{addr: $MACADDRESS, key: $AESKEY}

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 committed:
[server] Implement encryption/decryption

- A session struct is created in order to wrap the
  characteristic channel.
  It has an Encrypt method that takes a key and creates
  the AES/GCM cipher stored inside.
  This allows the protocol function to call sendMsg/recvMsg
  without caring of encryption itself: As long as the
  session.Encrypt method has been called, the session
  is encrypted.

- When the session is encrypted, each sent message is prepended
  with the IV used to encrypt it. Its structure is the following:

 0        4                                        16                                      Size + 4
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|  Size  |                   IV                   |  [Ciphertext [CBOR [Plaintext]]]...   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

- As several clients can be enrolled, different keys will
  be stored on server side, and we'll need to identify which
  client tries to run an attestation somehow.
  This is why a new message is sent by the client at the very
  beginning of the protocol, both at enroll and at attestation
  time.
  This message contains a client UUID, which allows to store the
  new key at enroll time, and to retrieve it during an attestation.
  The session.Encrypt method is called between the next message, so
  that except this UUID, everything is encrypted.

- The authentication message is moved before the enrollment one,
  because this message aims to avoid an attacker to trigger heavy TPM operations
  like key creation (mainly the Attestation Key), but the enrollment message
  actually creates one (or at least can).

- tpm functions are moved into a dedicated file

- RegistrationData structure is renamed to EnrollData, and its fields
  are also renamed to be more clear.

NOTE: The key is not yet persisted, thus only the enrollment is working

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 committed:
[server] Update deprecated architecture comment

The server architecture was reworked in previous
commits but the comment hasn't changed, leading to
outdated documentation of the server.

This commit updates this comment, providing an overview
of the server architecture, and a short description of
its key components.

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 committed:
[server] Add encryption key persistence

At enroll time, the key is saved in the form of a TPM sealed
object in two files:

- /etc/ultrablue/CLIENT_UUID
- /etc/ultrablue/CLIENT_UUID.pub

It is sealed to the Storage Root Key (SRK), so that if either the
TPM or the TPM owner is modified, those files couldn't be decrypted
anymore.

When running an attestation, the blobs are read from those files and
the resulting object is unsealed to retrieve the key.

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

nbouchinet-anssi commented on server/main.go:
```diff
@@ -20,6 +20,8 @@ var (
 	mtu          = flag.Int("mtu", 500, "Set a custom MTU, which is basically the max size of the BLE packets")
 	pcrextend    = flag.Bool("pcr-extend", false, "Extend the 9th PCR with the verifier secret on attestation success")
+
+	g_key []byte
```
If this isn't a command line argument, can you either change the previous comment or move this var in another var scope ?

kerneis-anssi replied:
And please add a comment, every global should be commented (except if it's self-describing, eg. command-line flags that have an embedded description string).

nbouchinet-anssi commented on server/main.go:
```diff
@@ -20,6 +20,8 @@ var (
 	mtu          = flag.Int("mtu", 500, "Set a custom MTU, which is basically the max size of the BLE packets")
 	pcrextend    = flag.Bool("pcr-extend", false, "Extend the 9th PCR with the verifier secret on attestation success")
+
+	g_key []byte
```
Is the `g_' prefix supposed to mean global ? If so, can you be consistent in your naming convention ?

kerneis-anssi commented:
In the commit message: "it's hex string" -> "its hex string"

kerneis-anssi commented:
Second commit message:
- It structures -> Its structure
- Use RFC-style representation for network packet.
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|    Type = 1   |    Length     |      MBZ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
```
(this one is copy-pasted from a random RFC but you get the idea)
- several client -> several clients
- it's field -> its field

See https://www.merriam-webster.com/words-at-play/when-to-use-its-vs-its for an explanation of it's vs its

e2r3p13 committed:
[server] Allow to provide a PIN to seal keys

Add a --with-pin option to the ultrablue server.
When used:

- At enroll time: The user is prompted to choose a PIN
to seal the encryption key on disk

- For an attestation: The PIN must be given
back to retrieve the encryption key

If the option is omitted for an attestation whereas it was
provided at enroll time, the policy will fail but no unsealing
attempt is made, thus the TPM fail counter is not incremented.

If the option is given for an attestation whereas it wasn't at
enroll time, the user will be prompted to enter a PIN, but any
PIN will do the job as it will not be used to unseal the key.

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

kerneis-anssi commented:
Third commit message: it's key components -> its key components.

nbouchinet-anssi commented on server/main.go:
```diff
@@ -84,6 +86,13 @@ func main() {
 
+	if *enroll {
+		logrus.Info("Generating symmetric key")
+		if g_key, err = TPM2_GetRandom(32); err != nil {
```
You seem to already use a splitted error checking syntax, could you use it everywhere ?

kerneis-anssi replied:
I agree it's better to be consistent, but keeping things on a single line is more compact and idiomatic go style. I'm not sure I'd have spotted the bug below with the split style for instance.

I suggest keeping this code as-is and a follow-up PR that would convert existing split style to single-line style.

What do you think?

e2r3p13 replied:
I personnally prefer the single-line version but there are some subtilities to consider because of how go assignation works:

```go
// The following creates somevar/err variables inside the if scope,
// meaning they doesn't change from the outer scope if they were declared.
if somevar, err := somefunc(); err != nil {
    ...
}

// This version updates outer variables, but they'll need to be declared before assignation
var somevar sometype
var err error

if somevar, err = somefunc(); err != nil {
    ...
}

// This one creates new variables but replaces previous ones with same name so that it's
// not necessary to declare them beforehand
err, somevar := somefunc()
if err != nil {
    ...
}
```
I agree however that it could be done in a follow-up commit that updates error handling with the chosen form

kerneis-anssi commented on server/protocol.go:
```diff
@@ -84,17 +32,17 @@ func recvMsg[T any](obj *T, ch chan []byte) error {
 // It also contains a boolean @PCRExtend that indicates the new verifier
 // it must generate a new secret to send back on attestation success.
-type RegistrationData struct {
-	Cert      []byte // x509 key certificate (one byte set to 0 if none)
```
It's good to use longer names, but please keep the comments, they are useful.

nbouchinet-anssi commented on server/main.go:
```diff
@@ -84,6 +86,13 @@ func main() {
 
+	if *enroll {
+		logrus.Info("Generating symmetric key")
+		if g_key, err = TPM2_GetRandom(32); err != nil {
```
Do `TPM2_GetRandom()` pledge that it will return a safe, non-null key even if no error is sent back ?

e2r3p13 replied:
I hope no null key can be returned without error but I didn't looked at the library code. I can check the key before returning it in my `TPM2_GetRandom` function to assert it is non null.

nbouchinet-anssi commented on server/main.go:
```diff
@@ -98,11 +107,12 @@ func main() {
 		logrus.Info("Generating enrollment QR code")
 		addr := device.Address().String()
-		qrcode, err := generateQRCode(addr)
+		json := fmt.Sprintf(`{"addr":"%s","key":"%x"}`, addr, g_key)
```
Could you use the [standard Golang json](https://pkg.go.dev/encoding/json) encoding library to build your json blob instead of a Sprintf ?

e2r3p13 replied:
I can have it working with `encoding/json` this way:
```go
qrcontent := struct {
	Addr, Key string
} {
	Addr: device.Address().String(),
	Key: fmt.Sprintf("%x", g_key),
}
qrjson, err := json.Marshal(qrcontent)
if err != nil {
	logrus.Fatal(err)
}
qrcode, err := generateQRCode(string(qrjson))
```
Which gives exactly the same result as:
```go
json := fmt.Sprintf(`{"addr":"%s","key":"%x"}`, addr, g_key)
qrcode, err := generateQRCode(json)
```
If you really prefer using the json standard library I can keep the former, but I would personally opt for the latter.

kerneis-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
+		close(ch)
+		return nil, err
+	}
+	if session.Encrypt(key); err != nil {
```
There is a bug here, it should be err = session.Encrypt(key) I guess?
Doesn't the go compiler warn about ignoring an error return type?

nbouchinet-anssi replied:
Agree, and be explicit even if it works.

e2r3p13 replied:
Yes, this is obviously a bug, and no, the go compiler doesn't complain at all.

kerneis-anssi reviewed (changes requested):
This is a review of the second commit. The first commit looks fine to me. I haven't looked at commits 3 and 4 yet.

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
+		return nil, err
+	}
+
+	if *enroll {
```
Could you remove this boilerplate code ?

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
+	} else {
+		// TODO: Fetch key from persistence point
+	}
+	if err != nil {
```
Duplicate of line 78 check.

Choose one of both but please be consistent in your choices.

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
-	b, err := getTPMRandom(16)
+func authentication(session *Session) error {
+	logrus.Info("Generating authentication nonce")
+	b, err := TPM2_GetRandom(16)
```
What `b' is supposed to be ? It seems to be a sensible part of your code, could you be explicit ?

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
+		close(session.ch)
 		return err
 	}
 	nonce := Bytestring{b}
```
I'd like you to explain here why do you cast a Byte array into your Bytestring type.

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
-	logrus.Info("Starting authentication process")
-	logrus.Info("Generating nonce")
-	b, err := getTPMRandom(16)
+func authentication(session *Session) error {
```
As we use AES/GCM, which I think operates as an authenticated encryption algorithm, we should be able to completely avoid this authentication step.

nbouchinet-anssi replied:
This code was originally implemented to prevent a malicious user from performing slow cryptographic operations such as tpm2_createak(). I don't know if we gain in performance in view of the TPM operations performed by this countermeasure.

However this sort of authentication adds an early anti-replay layer. We could keep it afterthought.

e2r3p13 replied:
We finally decided to remove the authentication message.

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -241,7 +219,7 @@ func response(ch chan []byte) error {
 	if len(response.Secret) > 0 {
 		logrus.Info("Extending PCR", PCR_EXTENSION_INDEX)
-		extendPCR(PCR_EXTENSION_INDEX, response.Secret)
+		TPM2_PCRExtend(PCR_EXTENSION_INDEX, response.Secret)
```
What if `TPM2_PCRExtend()` returns an error ?

nbouchinet-anssi commented on server/session.go:
```diff
@@ -0,0 +1,134 @@
+	return nil
+}
+
+func (s *Session) IsEncrypted() bool {
```
Unused ?

nbouchinet-anssi commented on server/session.go:
```diff
@@ -0,0 +1,134 @@
+		close(session.ch)
+		return err
+	}
+	if session.encrypted {
```
What's the use of your `IsEncrypted()` function ?

e2r3p13 replied:
Was for future use, but I think it never will.

nbouchinet-anssi commented on server/session.go:
```diff
@@ -0,0 +1,134 @@
+	}
+	if session.encrypted {
+		logrus.Debug("Encrypting (AES/GCM)")
+		iv, err := TPM2_GetRandom(IV_SIZE)
```
According to the The [crypto/cipher](https://pkg.go.dev/crypto/cipher) documentation, the nonce (your IV) must be NonceSize() bytes long.

e2r3p13 committed:
[server] Assert TPM2_GetRandom can't give null key

Checks the bytes returned by tpm2.GetRandom so that if it
silently fails without throwing any error, the TPM2_GetRandom
function will notice it and return one by itself.

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
-		- Each characteristic is declared in its own file, ending with Chr.go.
+	Ultrablue does the opposite. The server only exposes one characteristic, and
+	internally keeps the current state of the protocol for a particular connection.
+	Whenever the client reads or write on that characteristic, the server will
```
write -> writes

nbouchinet-anssi commented on server/session.go:
```diff
@@ -0,0 +1,134 @@
+			return err
+		}
+		ciphertext := session.aesgcm.Seal(nil, iv, data, nil)
+		data = append(iv, ciphertext...)
```
It seems that `Seal()` can append the nonce in its output could you check this ?

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
-		chunks in a for loop.
+	When the server is run, the characteristic is exposed, and it starts accepting
+	clients.
+	When a client connects and performs it's first operation on the characteristic, 
```
it's -> its

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	KEY COMPONENTS
+
+	- The ultrablue characteristic: When a client perform a read/write operation
+	handlers the characteristic will process it through handlers. It implements
```
Remove the first occurrence of "handlers"

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	---
+
+	KEY COMPONENTS
+
```
Each of the following components is implemented in a dedicated file named after the functionality.

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+
+	KEY COMPONENTS
+
+	- The ultrablue characteristic: When a client perform a read/write operation
```
perform -> performs

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	- The state: On each characteristic interaction, the "state" will be fetched
+	for that connection. It contains the bidirectional channel, a Session and
+	some internal fields. On the first client interaction, the state is created,
+	and the protocol is run.
```
is run -> starts running ?

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	an abstraction for other parts of the server, which is a bidirectional go
+	channel.
+
+	- The state: On each characteristic interaction, the "state" will be fetched
```
Start by saying what things are before saying how they behave:

every information that needs to be retained for the duration of a connection is stored in a State structure associated with a given client. The state is updated on each interaction with the characteristic. On the first ...

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	some internal fields. On the first client interaction, the state is created,
+	and the protocol is run.
+
+	- The protocol: Thanks to the control inversion provided by previously
```
control inversion -> inversion of control (I don't know why, but it's the term of art for this technique: https://en.wikipedia.org/wiki/Inversion_of_control)

kerneis-anssi replied:
Again, please explain first what this component does: orchestrates the successive stages of the protocol and transitions the state from one to the next?

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	described components, the protocol is defined in a linear way, this makes
+	things a lot easier to read and understand.
+
+	- The session: Created during the first steps of the protocol, it wraps the
```
Re-reading my previous review comments, I realize I don't understand from the content of this file alone what the difference is between the session and the state. I know that the Session is embedded in the State, but you should explain in this overview how they differ, what is the role of each one that makes it logical or necessary for them to be divided that way.

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+	- The session: Created during the first steps of the protocol, it wraps the
+	characteristic channel stored in the client state, and provides information
+	on how data is written to that channel.
+	The session is responsible to encrypt/decrypt data when needed.
```
If this is its only role, then it should come first in this comment: "the role of the session is to ..."

kerneis-anssi replied:
And maybe rename it to EncryptionSession then?

kerneis-anssi commented on server/main.go:
```diff
@@ -48,32 +48,55 @@ func initLogger(loglevel int) {
+
+	- The session: Created during the first steps of the protocol, it wraps the
+	characteristic channel stored in the client state, and provides information
+	on how data is written to that channel.
```
I don't understand this sentence, sorry. We can talk through it offline if it helps to refine the wording.

kerneis-anssi reviewed (changes requested):
Review of commit 3.

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -81,9 +81,14 @@ func establishEncryptedSession(ch chan []byte) (*Session, error) {
-		// TODO: Save key somewhere
+		key = make([]byte, 32)
+		copy(key, g_key)
+		g_key = nil
```
Do this safely invalidates the array ?

e2r3p13 replied:
It indicates to the go garbage collector that the memory is no longer used. As I don't know the go internals, I won't state that this safely invalidates the array. I would say that we should overwrite the key before releasing it.
As this could be useful at several places, I propose to let it as it for now and to create a function to erase/release memory later, and use it everywhere we need to.

kerneis-anssi commented on server/main.go:
```diff
@@ -22,6 +22,12 @@ var (
 	pcrextend    = flag.Bool("pcr-extend", false, "Extend the 9th PCR with the verifier secret on attestation success")
 )
 
+// Encryption key used at enroll time. It need to be globally available
```
It needs*

kerneis-anssi commented on server/protocol.go:
```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
+	if *enroll {
+		key = make([]byte, 32)
+		copy(key, enrollkey)
+		enrollkey = nil
```
This makes no sense to me. I remember some discussion about trying to clean up the key from memory, but I feel this little dance stems from a misunderstanding of Go's memory model. We should probably discuss at the whiteboard (and there should be a comment explaining what is going on if this is to be kept).

kerneis-anssi commented on server/session.go:
```diff
@@ -0,0 +1,127 @@
+// SPDX-FileCopyrightText: 2022 ANSSI
```
2023

kerneis-anssi commented on server/session.go:
```diff
@@ -0,0 +1,127 @@
+	be encrypted.
+*/
+func (s *Session) StartEncryption(key []byte) error {
+	block, err := aes.NewCipher(key)
```
First check that encryption is false and return an error otherwise.

kerneis-anssi commented on server/session.go:
```diff
@@ -0,0 +1,127 @@
+			close(session.ch)
+			return err
+		}
+		data = session.aesgcm.Seal(iv, iv, data, nil)
```
// Append encrypted data to the IV.

kerneis-anssi commented on server/session.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+/*
+	The Session type is an abstraction for the BLE go channel
```
"the BLE go channel" is unclear. Can you at least point to a file describing it, or better remind us of the context, eg.: "is an abstraction on top of the Go channel used to pass binary data to the goroutine reading and writing on the BLE characteristic. It takes care of automatically encoding and encrypting Go objects before sending them on the channel, and the other way around for received messages."

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -0,0 +1,47 @@
+	NOTE: It seems that it's not possible to add an event log
+	entry directly from the tss stack, and that we need to use
+	exposed UEFI function pointers.
+*/
```
Please document the restrictions on `secret` (I guess it must be a certain size that matches the size of the hashes for that particular PCR?).

e2r3p13 replied:
In fact, the only restriction is that the secret isn't `nil`. Even an empty string would work.

kerneis-anssi replied:
Ah, because you use PCREvent rather than PCRExtend. Fair enough. But then, is it really fine to just pass the secret as an Event, instead of a more structured binary blob? For instance, what happens the day we want to store this event in the event log, would it make a breaking change to all disks encrypted with ultrablue?

kerneis-anssi replied:
@nbouchinet-anssi may have more clues about the format of PCR Events.

kerneis-anssi replied:
Discussed offline, pushing this down the road: https://github.com/ANSSI-FR/ultrablue/issues/18

kerneis-anssi reviewed (changes requested):
Review of commits 1 and 2.

kerneis-anssi commented on server/main.go:
```diff
@@ -52,28 +52,53 @@ func initLogger(loglevel int) {
+	the server exposes characteristics and it is up to the client to read or
+	write them whenever it wants.
+
+	Ultrablue implements an inversion of control through serveral components,
```
serveral -> several

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+const SRK_HANDLE tpmutil.Handle = 0x81000001
+
+// TCG TPM v2.0 Provisioning Guidance, section 7.5.1:
+// https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
```
This section makes a reference to the default EK profile for RSA. Please extend the comment to also reference https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p4_r3.pdf Table B.3.3 (assuming that's the one you used).

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+
+/*
+	Returns a handle to the TPM Storage Rook Key (SRK) if it's already present
+	in it's non volatile memory.
```
it's -> its

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+/*
+	Returns a handle to the TPM Storage Rook Key (SRK) if it's already present
+	in it's non volatile memory.
+	Else, a new SRK is created using the default template, and is persisted
```
Else -> Otherwise

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+	Returns a handle to the TPM Storage Rook Key (SRK) if it's already present
+	in it's non volatile memory.
+	Else, a new SRK is created using the default template, and is persisted
+	in the TPM Non Volatile memory, at it's default location. A handle to the newly
```
it's -> its

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+	created key is then returned
+*/
+func TPM2_LoadSRK(rwc io.ReadWriteCloser) (tpmutil.Handle, error) {
+	handles, _, err := tpm2.GetCapability(rwc, tpm2.CapabilityHandles, 1000, uint32(tpm2.HandleTypePersistent) << 24)
```
Where do the count=1000 and <<24 come from?

e2r3p13 replied:
`1000` is an arbitrary value specifying the maximum object number `GetCapability` can return. I agree it deserves at least a const qualified name.

I don't remember where I found the `<< 24` trick, but I now I saw it somewhere (in a call to this same function). It is because depending on the capability given to `GetCapability`, the property argument can differ, but is an `uint32` whatever it means. For the `CapabilityHandle` capability, the tpm2 library provides constants, but they are `uint8`. The compiler enforces that an `uint32` is passed to the function, so I have either to cast or to convert the provided constant.

My guess is that when converting it to an uint32, the `uint8` value is implicitly shifted 24 bytes to the right (because intel machines are little endian), and that they cast the value back to an `uint8` inside the function, which results to a `0` if we don't shift it back.

If I'm right, an option would be to cast by myself the `uint8` to an `uint32`, instead of convert it. But casting isn't supported and I'd need dirty tricks as shown here: https://medium.com/a-journey-with-go/go-cast-vs-conversion-by-example-26e0ef3003f0

kerneis-anssi replied:
Let's add a comment to explain the situation:
// Shift the handle type into the MSO according to the specification: "TPM_CAP_HANDLES – Returns a list of all of the handles within the handle range of the property parameter. The range of the returned handles is determined by the handle type (the most-significant octet (MSO) of the property)."
// https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part3_Commands_pub.pdf Section 30.2 TPM2_GetCapability

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -4,10 +4,125 @@
+}
+
+/*
+	Tries to unseal the given object with the TPM Storage Root Key (SRK)
```
Remove "tries to" : every function can fail, this is already indicated by `error` in the return type.

kerneis-anssi commented on server/utils.go:
```diff
@@ -4,9 +4,64 @@
+	if err = os.MkdirAll(ULTRABLUE_KEYS_PATH, os.ModeDir); err != nil {
+		return err
+	}
+	if _, err = os.Stat(ULTRABLUE_KEYS_PATH + uuid); err == nil {
```
Note that there is a small race condition (TOCTOU). The proper way to do this atomically is os.Open(O_CREAT|O_EXCL) which will return an error if the file exists and create it otherwise. Let's assume though that we have a single process writing to those files at a time and not bother with it.

kerneis-anssi replied:
But maybe add a comment saying that we are aware of the race-condition and decide to ignore it?

e2r3p13 replied:
I prefer to fix this now as it will probably never be done otherwise.

kerneis-anssi reviewed (changes requested):
This is a review of the Seal/Unseal commit.

kerneis-anssi commented:
Meta comment: your commit messages are very informative. This information shouldn't only live in commit messages: can you copy-paste the informative parts (eg. documentation of --with-pin flag, structure of the IV+cypher packets, etc.) in either the relevant READMEs or some files under doc/. I don't really care where it lives, I'd just like to have more discoverable than in `git log`. All of those documentation updates can go in a single follow-up commit.

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -67,7 +67,7 @@ func TPM2_LoadSRK(rwc io.ReadWriteCloser) (tpmutil.Handle, error) {
 	returns the resulting private and public blobs
 */
-func TPM2_Seal(data []byte) ([]byte, []byte, error) {
+func TPM2_Seal(data []byte, pin string) ([]byte, []byte, error) {
```
Please document the new parameter. In particular, this function has a `pin` parameter but also looks at --with-pin: how are the two related?

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -84,10 +84,15 @@ func TPM2_Seal(data []byte) ([]byte, []byte, error) {
 	if sessHandle, _, err = tpm2.StartAuthSession(rwc, tpm2.HandleNull, tpm2.HandleNull, make([]byte, 16), nil, tpm2.SessionPolicy, tpm2.AlgNull, tpm2.AlgSHA256); err != nil {
 		return nil, nil, err
 	}
+	if *withpin {
```
Add a comment to explain the goal of this block. Idem in the next function.

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -84,10 +84,15 @@ func TPM2_Seal(data []byte) ([]byte, []byte, error) {
 	if sessHandle, _, err = tpm2.StartAuthSession(rwc, tpm2.HandleNull, tpm2.HandleNull, make([]byte, 16), nil, tpm2.SessionPolicy, tpm2.AlgNull, tpm2.AlgSHA256); err != nil {
 		return nil, nil, err
 	}
+	if *withpin {
```
Shouldn't we check instead that pin != "" ? Why do we care about the flag at this level of abstraction?

e2r3p13 replied:
pin can be an empty string in two different scenarios:
- The `withpin` flag isn't set
- The `withpin` flag is set but the user provided an empty string when asked

Depending on the presence of this option, and independently of the provided PIN, we must set a password policy or not:

If the `withpin` option is provided when unsealing, the user (or an attacker) will input a PIN. If it doesn't match, two things can happend based on how the enrollment was done:
- if the `withpin` option was provided: The policies now matches, but the PIN doesn't -> the DA counter is incremented and can lock the TPM
- if the `withpin` option wasn't provided: The policies doesn't match because it was a password policy at enroll time and it's not one at unesal time. The provided PIN isn't even tested, and the DA counter isn't incremented. The user can now restart the attestation without the `withpin` option

I don't know if this is clear, we can call to talk about it further. Anyway, I understand this isn't documented well enough, but I really don't know how to make these things clear in code comments, maybe it'll be easier to extract it in some README.md and pinpoint it in the code?

kerneis-anssi replied:
This makes sense to me. Please move the last paragraph of the function comment here, where it is easier to see what it refers to.

kerneis-anssi reviewed (changes requested):
This is a review of the PIN commit.

kerneis-anssi commented on server/tpm2.go:
```diff
@@ -147,6 +148,9 @@ func TPM2_GetRandom(size uint16) ([]byte, error) {
 		return nil, err
 	}
+	if rbytes == nil || len(rbytes) != int(size) || onlyContainsZeros(rbytes) {
+		return nil, errors.Errorf("Failed to generate random bytes")
```
Why not errors.New() like you did in another file?

e2r3p13 replied:
New is more adapted here. I'll change this.

kerneis-anssi reviewed (changes requested):
Minor comment about the last commit.

kerneis-anssi reviewed (dismissed):
A couple comments, but I'm fine with merging once you've addressed them.

kerneis-anssi reviewed (approved):
LGTM, assuming @nbouchinet-anssi is OK to resolve the remaining open comment about PCREvent.

kerneis-anssi commented on server/protocol.go:
```diff
@@ -10,65 +10,13 @@ import (
 	"os"
```
Please fix protocol test.

```
ultrablue/server$ go test
# ultrablue-server [ultrablue-server.test]
./protocol_test.go:27:23: cannot use ch (variable of type chan []byte) as type *Session in argument to sendMsg
./protocol_test.go:48:23: cannot use ch (variable of type chan []byte) as type *Session in argument to sendMsg
./protocol_test.go:93:24: cannot use ch (variable of type chan []byte) as type *Session in argument to recvMsg
./protocol_test.go:115:24: cannot use ch (variable of type chan []byte) as type *Session in argument to recvMsg
./protocol_test.go:132:24: cannot use ch (variable of type chan []byte) as type *Session in argument to recvMsg
FAIL	ultrablue-server [build failed]
```

kerneis-anssi reviewed (commented):
LGTM
